### ISSUE TITLE
convert_sync_batchnorm silently drops the activation / extra behavior of `_BatchNorm` subclasses (e.g. timm `BatchNormAct2d`) - please emit a warning

### ISSUE THREAD

**Issue by warptengood** [pytorch/pytorch#187298]
### 🐛 Describe the bug

`torch.nn.SyncBatchNorm.convert_sync_batchnorm()` rebuilds **every** `_BatchNorm` subclass as a *bare* `SyncBatchNorm`, copying only the BatchNorm attributes (running stats, weight, bias, eps, momentum). If the subclass added behavior - an overridden `forward`, a fused activation, dropout - that behavior is **silently discarded**. Child modules are even copied onto the new `SyncBatchNorm` (so `.act` is still *present*), but `SyncBatchNorm.forward` never calls them, so they become dead attributes. No error, no warning.

The result is a model that **computes a different function after conversion than before it**, with nothing to alert the user.

### Minimal, self-contained repro (pure PyTorch, CPU, no distributed init)

```python
import torch
import torch.nn as nn

# A BatchNorm subclass that fuses an activation after normalization -
# exactly the shape of timm's BatchNormAct2d (pervasive in timm encoders).
class BatchNormAct2d(nn.BatchNorm2d):
    def __init__(self, num_features):
        super().__init__(num_features)
        self.act = nn.ReLU()
    def forward(self, x):
        return self.act(super().forward(x))   # normalize -> activation

torch.manual_seed(0)
model = nn.Sequential(BatchNormAct2d(4)).eval()
x = torch.randn(2, 4, 8, 8)
before = model(x)

converted = nn.SyncBatchNorm.convert_sync_batchnorm(model).eval()
after = converted(x)

before = before.detach()
after  = after.detach()

print("converted submodule type :", type(converted[0]).__name__)
print("ReLU applied before convert (min>=0):", float(before.min()) >= 0)
print("ReLU applied after  convert (min>=0):", float(after.min())  >= 0)
print("outputs differ after convert        :", not torch.allclose(before, after))
print("orphaned .act still present, unused  :", hasattr(converted[0], "act"))
print("max abs diff                         :", float((before - after).abs().max()))
```

Output (torch 2.11.0):

```
converted submodule type : SyncBatchNorm
ReLU applied before convert (min>=0): True
ReLU applied after  convert (min>=0): False  # <-- activation gone, negatives pass through
outputs differ after convert        : True
orphaned .act still present, unused : True  # <-- ReLU copied onto SyncBatchNorm but never called
max abs diff                        : 3.1537086963653564
```

The activation is dropped, yet `.act` is still attached to the module - it's just never invoked. Nothing warns.

This reproduces unchanged on the latest nightly (`2.14.0.dev20260614`), not just the release build below.

### Real-world impact

This is the single most common multi-GPU setup for any [timm](https://github.com/huggingface/pytorch-image-models)-based model. timm's encoders use `BatchNormAct2d` (fused BN + activation) *pervasively* - e.g. a MobileNetV4 encoder has **36 fused ReLUs**. The instant you wrap for DDP with `torch.nn.SyncBatchNorm.convert_sync_batchnorm`, all 36 activations vanish.

The failure mode is maximally nasty because **it is silent and deferred**:
- The model trains fine on the (silently degenerate) graph and reports a healthy validation metric.
- At inference / ONNX-export time the model is rebuilt from the *original* module definitions (activations present), the trained weights are loaded into a now-different graph, and the model collapses - in our case a segmentation model that logged **0.82 val F1 exported to 0.00** (all-background predictions).
- We burned ~2.6 days on 4 GPUs before realizing the trained graph and the exported graph were different networks. The clue was only visible by diffing module *types* after conversion (`{SyncBatchNorm}` vs `{BatchNorm2d, BatchNormAct2d}`).

A single warning at conversion time would have made this a five-minute fix instead of a multi-day investigation.

### Other libraries already work around this

- **timm** has had this reported since 2022 and ships its own fix - [huggingface/pytorch-image-models#1254 "[BUG] Broken SyncBatchNorm due to BatchNormAct2d"](https://github.com/huggingface/pytorch-image-models/issues/1254). Their EfficientNet repro went from a sane output to `-1.4e17` after conversion. Fix: a custom [`SyncBatchNormAct`](https://github.com/huggingface/pytorch-image-models/blob/fbe27d6c935a00e357e13f2f74043645769bd5fc/timm/layers/norm_act.py#L167) + [`timm.layers.convert_sync_batchnorm`](https://github.com/huggingface/pytorch-image-models/blob/main/timm/utils/model.py) that re-applies `.act`/`.drop`.
- **mmengine** ships its own [`convert_sync_batchnorm`](https://mmengine.readthedocs.io/en/latest/api/generated/mmengine.model.convert_sync_batchnorm.html).
- **detectron2** ships its own conversion [helpers](https://github.com/facebookresearch/detectron2/blob/main/detectron2/layers/batch_norm.py) as well.

Three major libraries independently reimplemented this function specifically to route around the stock one. That's a strong signal the stock converter has a sharp edge worth guarding.

### Precedent: PyTorch already treated this `isinstance(module, _BatchNorm)` overreach as a bug

`convert_sync_batchnorm` selects layers with `isinstance(module, _BatchNorm)`. That check *used to* also catch `nn.InstanceNorm*`, because `_InstanceNorm` inherited from `_BatchNorm` - so instance-norm layers were silently converted to `SyncBatchNorm` and broken. PyTorch fixed this in commit [`37ca5a8`](https://github.com/pytorch/pytorch/commit/37ca5a8a644a21efcb1bf37c1da66ef81133af06) - *"convert_sync_batchnorm should not convert _InstanceNorm instances"* - by introducing a shared `_NormBase` base class and **reparenting `_InstanceNorm` out from under `_BatchNorm`**, so the `isinstance(module, _BatchNorm)` check no longer matches it.

So the project already recognizes this converter wrongly swallowing `_BatchNorm`-matching layers as a real defect. That fix worked because `_InstanceNorm` is in-tree and could be reparented. External subclasses like timm's `BatchNormAct2d` legitimately **are** `BatchNorm` layers (they genuinely want the synced BN) and cannot be reparented out - so the only general remedy is to **warn** when extra behavior is about to be silently dropped.

### Expected behavior / proposed fix

I'm **not** asking PyTorch to magically preserve arbitrary subclass behavior - that's not knowable in general. Just make the silent data-loss loud:

> When `convert_sync_batchnorm` is about to rebuild a `module` whose `type(module)` is a strict subclass of `_BatchNorm` that **overrides `forward`** and/or **has child modules / extra registered attributes**, emit a `warnings.warn(...)` (or `UserWarning`) saying that custom behavior will not be carried over to the bare `SyncBatchNorm` and may silently change the model's output.

This is cheap, non-breaking, and would have caught the timm/mmengine/detectron2 footgun at the source. A heuristic like `type(module).forward is not _BatchNorm.forward` or `len(list(module.children())) > 0` would cover the real cases.

Happy to send a PR for the warning if maintainers agree on the approach.

### Versions

```bash
PyTorch version: 2.11.0+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.0 (main, Oct  3 2023, 01:27:23) [Clang 17.0.1 ] (64-bit runtime)
Python platform: Linux-6.8.0-124-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.0.140
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 2080 Ti
GPU 1: NVIDIA GeForce RTX 2080 Ti
GPU 2: NVIDIA GeForce RTX 2080 Ti
GPU 3: NVIDIA GeForce RTX 2080 Ti

Nvidia driver version: 570.211.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.2
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  32
On-line CPU(s) list:                     0-31
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Silver 4110 CPU @ 2.10GHz
CPU family:                              6
Model:                                   85
Thread(s) per core:                      2
Core(s) per socket:                      8
Socket(s):                               2
Stepping:                                4
CPU(s) scaling MHz:                      41%
CPU max MHz:                             3000.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4200.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 pti intel_ppin ssbd mba ibrs ibpb stibp tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts vnmi pku ospke md_clear flush_l1d arch_capabilities ibpb_exit_to_user
Virtualization:                          VT-x
L1d cache:                               512 KiB (16 instances)
L1i cache:                               512 KiB (16 instances)
L2 cache:                                16 MiB (16 instances)
L3 cache:                                22 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-7,16-23
NUMA node1 CPU(s):                       8-15,24-31
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             KVM: Mitigation: VMX disabled
Vulnerability L1tf:                      Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:                       Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:                  Mitigation; PTI
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Mitigation; IBRS
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; IBRS; IBPB conditional; STIBP conditional; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect
```

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy
